sklearn/manifold/isomap.py

```diff
@@ -197,17 +211,20 @@ def transform(self, X):
 
         Parameters
         ----------
-        X : array-like, shape (n_samples, n_features)
+        X : array-like, shape (n_samples_transform, n_features)
+            If neighbors_algorithm='precomputed', X is assumed to be a
+            distance matrix of shape (n_samples_transform, n_samples_fit).
 
         Returns
         -------
-        X_new : array-like, shape (n_samples, n_components)
+        X_new : array-like, shape (n_samples_transform, n_components)
         """
-        X = check_array(X)
+        check_is_fitted(self, 'embedding_')
+        X = check_array(X, accept_sparse=['csr'])
         distances, indices = self.nbrs_.kneighbors(X, return_distance=True)
 
-        # Create the graph of shortest distances from X to self.training_data_
-        # via the nearest neighbors of X.
+        # Create the graph of shortest distances from X to
+        # self.training_data_ via the nearest neighbors of X.
```

Comment:
If `training_data_` is only used for its shape, I think it should be deprecated and removed.